msg.c

```diff
@@ -26,136 +26,130 @@
 
 static u32 msg_csum(msg_t *msg)
 {
-    u32 k;
-    u16 csum;
-
-    csum = 0;
-    for (k = 0; k < MSG_HDRSZ + msg_getlen(msg); k++) {
-        if (k >= sizeof(msg->csum))
-            csum += *((u8 *)msg + k);
-    }
-    csum += msg_getseq(msg);
-    return csum;
+	u32 k;
+	u16 csum;
+
+	csum = 0;
+	for (k = 0; k < MSG_HDRSZ + msg_getlen(msg); k++) {
+		if (k >= sizeof(msg->csum))
+			csum += *((u8 *)msg + k);
+	}
+	csum += msg_getseq(msg);
+	return csum;
 }
 
 
-int msg_safewrite(u16 pn, u8 *buff, u16 len)
+static int msg_write(u16 pn, msg_t *msg)
 {
-    int l;
-
-    for (l = 0; len;) {
-        if ((l = serial_write(pn, buff, len)) < 0)
-            return ERR_MSG_IO;
-        buff += l;
-        len -= l;
-    }
-    return 0;
+	u8 *p = (u8 *)msg;
+	char cs[2];
+	u16 k;
+	int res;
+
+	/* Frame start */
+	cs[0] = MSG_MARK;
+	if ((res = msg->write(pn, (u8 *)cs, 1)) < 0)
+		return res;
+
+	for (k = 0; k < MSG_HDRSZ + msg_getlen(msg); k++) {
+		if ((p[k] == MSG_MARK) || (p[k] == MSG_ESC)) {
+			cs[0] = MSG_ESC;
+			if (p[k] == MSG_MARK)
+				cs[1] = MSG_ESCMARK;
+			else
+				cs[1] = MSG_ESCESC;
+			if ((res = msg->write(pn, (u8 *)cs, 2)) < 0)
+				return res;
+		}
+		else {
+			if ((res = msg->write(pn, (u8 *)&p[k], 1)) < 0)
+				return res;
+		}
+	}
```

Comment:
```suggestion
	u8 cs[2];
	u16 k;
	int res;

	/* Frame start */
	cs[0] = MSG_MARK;
	if ((res = msg->write(pn, cs, 1)) < 0)
		return res;

	for (k = 0; k < MSG_HDRSZ + msg_getlen(msg); k++) {
		if ((p[k] == MSG_MARK) || (p[k] == MSG_ESC)) {
			cs[0] = MSG_ESC;
			cs[1] = p[k] == MSG_MARK ? MSG_ESCMARK : MSG_ESCESC;
			if ((res = msg->write(pn, cs, 2)) < 0)
				return res;
		}
		else {
			if ((res = msg->write(pn, &p[k], 1)) < 0)
				return res;
		}
	}
```
And also by the way:
- `cs`  as `u8` type instead of `char`
- `pk` is of type `u8 *`

Reply:
The same as above.